File: PHP/leetcode/shift-2d-grid/solution.php
class Solution {

    /**
     * @param Integer[][] $grid
     * @param Integer $k
     * @return Integer[][]
     */
    function shiftGrid($grid, $k) {
        $m = count($grid);
        if ($m == 0) return $grid;
        $n = is_array($grid[0]) ? count($grid[0]) : 0;
        if ($n == 0) return $grid;
        $total = $m * $n;
        $k %= $total;
        $flat = [];
        for ($i = 0; $i < $m; $i++)
            for ($j = 0; $j < $n; $j++)
                $flat[] = $grid[$i][$j];
        $res = [];
        for ($i = 0; $i < $m; $i++) $res[$i] = array_fill(0, $n, 0);
        for ($idx = 0; $idx < $total; $idx++) {
            $np = ($idx + $k) % $total;
            $res[intdiv($np, $n)][$np % $n] = $flat[$idx];
        }
        return $res;
    }
}

File: Swift/leetcode/shift-2d-grid/solution.swift
class Solution {
    func shiftGrid(_ grid: [[Int]], _ k: Int) -> [[Int]] {
        let m = grid.count, n = grid[0].count, total = m*n, kk = k % total
        var res = Array(repeating: Array(repeating: 0, count: n), count: m)
        for i in 0..<total { let ni = (i + kk) % total; res[ni/n][ni%n] = grid[i/n][i%n] }
        return res
    }
}
